File: app/src/main/java/org/ole/planet/myplanet/ui/library/TagExpandableAdapter.java
package org.ole.planet.myplanet.ui.library;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseExpandableListAdapter;
import android.widget.CheckBox;

import androidx.appcompat.widget.AppCompatImageView;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.databinding.RowAdapterNavigationChildBinding;
import org.ole.planet.myplanet.databinding.RowAdapterNavigationParentBinding;
import org.ole.planet.myplanet.model.RealmTag;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class TagExpandableAdapter extends BaseExpandableListAdapter {
    private Context context;
    private List<RealmTag> tagList;
    private OnClickTagItem clickListener;
    private boolean isSelectMultiple = false;
    private ArrayList<RealmTag> selectedItemsList = new ArrayList<>();
    private HashMap<String, List<RealmTag>> childMap;
    public TagExpandableAdapter(Context context, List<RealmTag> tagList, HashMap<String, List<RealmTag>> childMap, ArrayList<RealmTag> selectedItemsList) {
        this.context = context;
        this.tagList = tagList;
        this.childMap = childMap;
        this.selectedItemsList = selectedItemsList;
    }

    public void setSelectMultiple(boolean selectMultiple) {
        isSelectMultiple = selectMultiple;

    }

    @Override
    public int getGroupCount() {
        return tagList.size();
    }

    @Override
    public int getChildrenCount(int groupPosition) {
        if (childMap.containsKey(tagList.get(groupPosition).getId())) {
            return childMap.get(tagList.get(groupPosition).getId()).size();
        }
        return 0;
    }

    @Override
    public Object getGroup(int groupPosition) {
        return tagList.get(groupPosition);
    }

    @Override
    public Object getChild(int groupPosition, int childPosition) {
        if (childMap.containsKey(tagList.get(groupPosition).getId())) {
            return childMap.get(tagList.get(groupPosition).getId()).get(childPosition);
        }
        return null;
    }

    @Override
    public long getGroupId(int groupPosition) {
        return groupPosition;
    }

    @Override
    public long getChildId(int groupPosition, int childPosition) {
        return childPosition;
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }

    @Override
    public View getGroupView(final int groupPosition, final boolean isExpanded, View convertView, ViewGroup parent) {
        String headerTitle = tagList.get(groupPosition).getName();

        RowAdapterNavigationParentBinding rowAdapterNavigationParentBinding;
        if (convertView == null) {
            rowAdapterNavigationParentBinding = RowAdapterNavigationParentBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            convertView = rowAdapterNavigationParentBinding.getRoot();
            convertView.setTag(rowAdapterNavigationParentBinding);
        } else {
            rowAdapterNavigationParentBinding = (RowAdapterNavigationParentBinding) convertView.getTag();
        }

        rowAdapterNavigationParentBinding.tvDrawerTitle1.setText(headerTitle);
        createCheckbox(convertView, tagList.get(groupPosition));
        rowAdapterNavigationParentBinding.tvDrawerTitle.setText(headerTitle);

        if (!childMap.containsKey(tagList.get(groupPosition).getId())) {
            rowAdapterNavigationParentBinding.tvDrawerTitle1.setVisibility(View.VISIBLE);
            rowAdapterNavigationParentBinding.tvDrawerTitle.setVisibility(View.GONE);
            rowAdapterNavigationParentBinding.ivIndicators.setVisibility(View.GONE);
            rowAdapterNavigationParentBinding.tvDrawerTitle1.setOnClickListener(v -> clickListener.onTagClicked(tagList.get(groupPosition)));
        } else {
            rowAdapterNavigationParentBinding.tvDrawerTitle.setVisibility(View.VISIBLE);
            rowAdapterNavigationParentBinding.tvDrawerTitle1.setOnClickListener(null);
            rowAdapterNavigationParentBinding.tvDrawerTitle1.setVisibility(View.GONE);
            rowAdapterNavigationParentBinding.ivIndicators.setVisibility(View.VISIBLE);
            setExpandedIcon(isExpanded, rowAdapterNavigationParentBinding.ivIndicators);
            rowAdapterNavigationParentBinding.tvDrawerTitle.setOnClickListener(v -> clickListener.onTagClicked(tagList.get(groupPosition)));
        }

        return convertView;
    }

    private void setExpandedIcon(boolean isExpanded, AppCompatImageView ivIndicator) {
        if (isExpanded) {
            ivIndicator.setImageResource(R.drawable.ic_keyboard_arrow_up_black_24dp);
        } else {
            ivIndicator.setImageResource(R.drawable.ic_keyboard_arrow_down_black_24dp);
        }
    }

    private View inflateView(ViewGroup parent, int layout) {
        LayoutInflater inflater = LayoutInflater.from(context);
        return inflater.inflate(layout, parent, false);
    }

    private void createCheckbox(View convertView, RealmTag tag) {
        CheckBox checkBox = convertView.findViewById(R.id.checkbox);
        checkBox.setVisibility(isSelectMultiple ? View.VISIBLE : View.GONE);
        checkBox.setChecked(selectedItemsList.contains(tag));
        checkBox.setOnCheckedChangeListener((compoundButton, b) -> {
            clickListener.onCheckboxTagSelected(tag);
        });
    }

    @Override
    public View getChildView(final int groupPosition, final int childPosition, boolean isLastChild, View convertView, ViewGroup parent) {
        RealmTag tag = (RealmTag) getChild(groupPosition, childPosition);

        RowAdapterNavigationChildBinding rowAdapterNavigationChildBinding;
        if (convertView == null) {
            rowAdapterNavigationChildBinding = RowAdapterNavigationChildBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            convertView = rowAdapterNavigationChildBinding.getRoot();
            convertView.setTag(rowAdapterNavigationChildBinding);
        } else {
            rowAdapterNavigationChildBinding = (RowAdapterNavigationChildBinding) convertView.getTag();
        }

        createCheckbox(convertView, tag);
        rowAdapterNavigationChildBinding.tvDrawerTitle.setText(tag.getName());
        rowAdapterNavigationChildBinding.tvDrawerTitle.setOnClickListener(v -> {
            if (clickListener != null) {
                clickListener.onTagClicked(tag);
            }
        });

        return convertView;
    }

    @Override
    public boolean isChildSelectable(int groupPosition, int childPosition) {
        return false;
    }

    public void setClickListener(OnClickTagItem clickListener) {
        this.clickListener = clickListener;
    }

    public void setTagList(List<RealmTag> filteredList) {
        this.tagList = filteredList;
        notifyDataSetChanged();
    }

    public interface OnClickTagItem {
        void onTagClicked(RealmTag tag);

        void onCheckboxTagSelected(RealmTag tags);
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/library/TagExpandableAdapter.kt
package org.ole.planet.myplanet.ui.library

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseExpandableListAdapter
import android.widget.CheckBox
import android.widget.CompoundButton
import androidx.appcompat.widget.AppCompatImageView
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.RowAdapterNavigationChildBinding
import org.ole.planet.myplanet.databinding.RowAdapterNavigationParentBinding
import org.ole.planet.myplanet.model.RealmTag

class TagExpandableAdapter(private var tagList: List<RealmTag>, private val childMap: HashMap<String?, MutableList<RealmTag>>, selectedItemsList: MutableList<RealmTag>) : BaseExpandableListAdapter() {
    private var clickListener: OnClickTagItem? = null
    private var isSelectMultiple = false
    private var selectedItemsList = ArrayList<RealmTag>()

    init {
        this.selectedItemsList = selectedItemsList as ArrayList<RealmTag>
    }

    fun setSelectMultiple(selectMultiple: Boolean) {
        isSelectMultiple = selectMultiple
    }

    override fun getGroupCount(): Int {
        return tagList.size
    }

    override fun getChildrenCount(groupPosition: Int): Int {
        return if (childMap.containsKey(tagList[groupPosition].id)) {
            childMap[tagList[groupPosition].id]!!.size
        } else 0
    }

    override fun getGroup(groupPosition: Int): Any {
        return tagList[groupPosition]
    }

    override fun getChild(groupPosition: Int, childPosition: Int): RealmTag? {
        return if (childMap.containsKey(tagList[groupPosition].id)) {
            childMap[tagList[groupPosition].id]!![childPosition]
        } else null
    }

    override fun getGroupId(groupPosition: Int): Long {
        return groupPosition.toLong()
    }

    override fun getChildId(groupPosition: Int, childPosition: Int): Long {
        return childPosition.toLong()
    }

    override fun hasStableIds(): Boolean {
        return false
    }

    override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View, parent: ViewGroup): View {
        var convertView: View? = convertView
        val headerTitle = tagList[groupPosition].name
        val rowAdapterNavigationParentBinding: RowAdapterNavigationParentBinding
        if (convertView == null) {
            rowAdapterNavigationParentBinding = RowAdapterNavigationParentBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            convertView = rowAdapterNavigationParentBinding.root
            convertView.setTag(rowAdapterNavigationParentBinding)
        } else {
            rowAdapterNavigationParentBinding = convertView.tag as RowAdapterNavigationParentBinding
        }
        rowAdapterNavigationParentBinding.tvDrawerTitle1.text = headerTitle
        createCheckbox(convertView, tagList[groupPosition])
        rowAdapterNavigationParentBinding.tvDrawerTitle.text = headerTitle
        if (!childMap.containsKey(tagList[groupPosition].id)) {
            rowAdapterNavigationParentBinding.tvDrawerTitle1.visibility = View.VISIBLE
            rowAdapterNavigationParentBinding.tvDrawerTitle.visibility = View.GONE
            rowAdapterNavigationParentBinding.ivIndicators.visibility = View.GONE
            rowAdapterNavigationParentBinding.tvDrawerTitle1.setOnClickListener {
                clickListener!!.onTagClicked(tagList[groupPosition])
            }
        } else {
            rowAdapterNavigationParentBinding.tvDrawerTitle.visibility = View.VISIBLE
            rowAdapterNavigationParentBinding.tvDrawerTitle1.setOnClickListener(null)
            rowAdapterNavigationParentBinding.tvDrawerTitle1.visibility = View.GONE
            rowAdapterNavigationParentBinding.ivIndicators.visibility = View.VISIBLE
            setExpandedIcon(isExpanded, rowAdapterNavigationParentBinding.ivIndicators)
            rowAdapterNavigationParentBinding.tvDrawerTitle.setOnClickListener {
                clickListener!!.onTagClicked(tagList[groupPosition])
            }
        }
        return convertView
    }

    private fun setExpandedIcon(isExpanded: Boolean, ivIndicator: AppCompatImageView) {
        if (isExpanded) {
            ivIndicator.setImageResource(R.drawable.ic_keyboard_arrow_up_black_24dp)
        } else {
            ivIndicator.setImageResource(R.drawable.ic_keyboard_arrow_down_black_24dp)
        }
    }

    private fun createCheckbox(convertView: View, tag: RealmTag) {
        val checkBox = convertView.findViewById<CheckBox>(R.id.checkbox)
        checkBox.visibility = if (isSelectMultiple) View.VISIBLE else View.GONE
        checkBox.isChecked = selectedItemsList.contains(tag)
        checkBox.setOnCheckedChangeListener { _: CompoundButton?, _: Boolean ->
            clickListener!!.onCheckboxTagSelected(tag)
        }
    }

    override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convertView: View, parent: ViewGroup): View {
        var convertView: View? = convertView
        val tag = getChild(groupPosition, childPosition) as RealmTag
        val rowAdapterNavigationChildBinding: RowAdapterNavigationChildBinding
        if (convertView == null) {
            rowAdapterNavigationChildBinding = RowAdapterNavigationChildBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            convertView = rowAdapterNavigationChildBinding.root
            convertView.setTag(rowAdapterNavigationChildBinding)
        } else {
            rowAdapterNavigationChildBinding = convertView.tag as RowAdapterNavigationChildBinding
        }
        createCheckbox(convertView, tag)
        rowAdapterNavigationChildBinding.tvDrawerTitle.text = tag.name
        rowAdapterNavigationChildBinding.tvDrawerTitle.setOnClickListener {
            if (clickListener != null) {
                clickListener!!.onTagClicked(tag)
            }
        }
        return convertView
    }

    override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
        return false
    }

    fun setClickListener(clickListener: OnClickTagItem?) {
        this.clickListener = clickListener
    }

    fun setTagList(filteredList: List<RealmTag>) {
        tagList = filteredList
        notifyDataSetChanged()
    }

    interface OnClickTagItem {
        fun onTagClicked(tag: RealmTag)
        fun onCheckboxTagSelected(tags: RealmTag)
    }
}
